src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs

```diff
@@ -0,0 +1,536 @@
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(consensusTip), consensusTip);
+
+            ChainedHeader current = consensusTip;
+            while (current.Previous != null)
+            {
+                current.Previous.Next.Add(current);
+                this.chainedHeadersByHash.Add(current.HashBlock, current);
+                current = current.Previous;
+            }
+
+            if (current.HashBlock != this.network.GenesisHash)
+            {
+                this.logger.LogTrace("(-)[INVALID_NETWORK]");
+                throw new ConsensusException("INVALID_NETWORK");
+            }
+
+            this.SetPeerTip(LocalPeerId, consensusTip.HashBlock);
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// A new list of headers are presented from a given peer, the headers will try to be connected to the chain.
+        /// Headers that are interesting (i.e may extend our consensus tip) will marked to download there full blocks.
+        /// </summary>
+        /// <remarks>
+        /// The headers are assumed to be consecutive in order. 
+        /// </remarks>
+        /// <param name="networkPeerId">The network id that is presenting this headers.</param>
+        /// <param name="headers">The list of headers to connect to the chain tree.</param>
+        /// <returns>Indicators of what blocks need to be downloaded.</returns>
+        public ConnectedHeaders ConnectNewHeaders(int networkPeerId, List<BlockHeader> headers)
+        {
+            Guard.NotNull(headers, nameof(headers));
+            this.logger.LogTrace("({0}:{1},{2}.{3}:{4})", nameof(networkPeerId), networkPeerId, nameof(headers), nameof(headers.Count), headers.Count);
+
+            if (!this.chainedHeadersByHash.ContainsKey(headers[0].HashPrevBlock))
+            {
+                this.logger.LogTrace("(-)[HEADER_COULD_NOT_CONNECT]");
+                throw new ConnectHeaderException();
+            }
+
+            List<ChainedHeader> newChainedHeaders = this.CreateNewHeaders(headers);
+
+            this.SetPeerTip(networkPeerId, headers.Last().GetHash());
+
+            if (newChainedHeaders == null)
+            {
+                this.logger.LogTrace("(-)[NO_NEW_HEADER]");
+                return new ConnectedHeaders();
+            }
+
+            ChainedHeader earliestNewHeader = newChainedHeaders.First();
+            ChainedHeader latestNewHeader = newChainedHeaders.Last();
+
+            ConnectedHeaders connectedHeaders = null;
+
+            bool isAssumedValidEnabled = this.consensusSettings.BlockAssumedValid != null;
+            bool isWithinCheckpoints = this.consensusSettings.UseCheckpoints && (earliestNewHeader.Height <= this.checkpoints.GetLastCheckpointHeight()); ;
+
+            if (isWithinCheckpoints || isAssumedValidEnabled)
+            {
+                ChainedHeader currentChainedHeader = latestNewHeader;
+
+                while (currentChainedHeader != earliestNewHeader)
```

Comment:
Please add a comment about going in backwards direction and why we need to do so

Reply:
Agree

When we look for a checkpoint header or an assume valid header, we go from the last presented header to the first one in the reverse order, because if there were multiple checkpoints or a checkpoint and an assume valid header inside of the presented list of headers, we would be interested only in the last one as it would cover all previous headers. Reversing the order of processing guarantees that we only need to deal with one special header, which simplifies the implementation.